[esphome] Fix OTA watchdog resets during port scanning and network delays

bdraco opened the pull request:
# What does this implement/fix?

This PR fixes two critical bugs that cause ESP devices to reset due to watchdog timeout:

1. **Port scanning reset**: When the OTA port (3232 on ESP32, 8266 on ESP8266) is scanned by tools like nmap or when any client connects without sending data
2. **Network delay reset**: When the OTA client stalls sending data for more than ~5 seconds during an update, triggering the watchdog

The root cause was that the OTA component never set the client socket to non-blocking mode after accepting connections, despite having proper non-blocking I/O handling code in place. This caused the device to block indefinitely waiting for data, triggering the watchdog timer after ~5 seconds.

**This bug has existed since the initial implementation of the OTA component** - the code was written with non-blocking socket semantics (checking for `EAGAIN`/`EWOULDBLOCK` errors in `readall_()` and `writeall_()`), but the client socket was never actually set to non-blocking mode. This oversight meant the non-blocking code paths were never executed.

Additionally, this PR improves security by deferring OTA status callbacks and warning flags until after successful authentication, preventing port scanners from triggering false OTA update notifications.

**Major architectural changes:**
- Split OTA handling into non-blocking `handle_handshake_()` and blocking `handle_data_()` phases
- Implemented proper timeout handling with 10-second handshake timeout and 90-second timeout per read/write operation
- Used `accept_loop_monitored()` for better event loop integration (matching API server pattern)

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- fixes #<issue number> (Discord report from chorÐ¢нørd on 7/17/25 about nmap causing ESP resets)

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- Not applicable (internal fix, no user-facing changes)

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [x] ESP8266
- [ ] RP2040
- [x] BK72xx
- [ ] RTL87xx
- [ ] nRF52840

## Example entry for `config.yaml`:

```yaml
# No configuration changes required
# This fix applies to all existing OTA configurations:
ota:
  - platform: esphome
    password: "your_ota_password"
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

## Additional Information

### Changes made:
1. **Set client socket to non-blocking mode** after accept() to prevent blocking reads
2. **Split OTA handling** into non-blocking handshake phase and data transfer phase
3. **Implemented dual timeout system**: 10 seconds for initial handshake, 90 seconds per read/write operation during data transfer
4. **Moved timeout constants to header file** for better maintainability
5. **Use `accept_loop_monitored()`** instead of regular `accept()` for better event loop integration
6. **Added connection time tracking** to enforce handshake timeout
7. **Moved OTA callbacks** to execute after authentication instead of on connection

### Testing performed:
- Verified `nmap -p 3232 -T4 -A -v <esp32_ip>` no longer causes watchdog reset
- Confirmed OTA updates still work correctly with esphome CLI
- Tested with password-protected and non-password OTA configurations
- Verified port scanners no longer trigger OTA status callbacks
- Successfully performed OTA update over VPN to a device 4000 miles away with high latency
- Updated timeouts to 10s (handshake) and 90s (per read/write operation) based on community feedback to better support marginal network conditions
- The 90s timeout matches the timeout currently used in aioesphomeapi for consistency across ESPHome communication protocols

### Performance impact:
- **Flash usage**: -56 bytes (actually reduced!) on ESP32
- **RAM usage**: No increase (+4 bytes if alignment doesn't work out on some platforms)
- **Runtime performance**: Improved - non-blocking handshake phase prevents port scanners from blocking the main loop

bdraco committed:
[esphome] Fix OTA watchdog reset when port scanning

RubenKelevra commented:
Damn, that could explain also my issues with updates of devices far away from the AP, where bursts of bad reception seem to happen (probably due to noise in the 2.4 GHz sprectrum).

RubenKelevra commented:
I’ve quickly read over the PR. If I read it correctly, it will abort the OTA connection if no bytes are received for 2.5 s. This is a major issue in practice, because on marginal 2.4 GHz links **the 5 s watchdog already trips during ordinary OTA updates**: the device stops reading for a few seconds, then reboots mid‑transfer. In captures I see long gaps with no data, followed by the node advertising via mDNS; the next retransmit from the host gets ACKed with RST, which cleanly matches a reboot.

What the packet trace shows on ESP32‑C3 with ESPHome’s default ESP‑IDF/lwIP config: SACK is off and TCP window scaling is off, so the ESP advertises a small fixed receive window (~4 752 B). With a typical MSS ≈ 1 460 B, that leaves ≈3 segments in flight. Under even mild loss, recovery via duplicate ACKs is often impossible; we fall into RTO backoff instead. That explains the multi‑second and sometimes multi‑minute “no progress” gaps despite a live TCP session. A single RTO at these throughputs is enough to exceed the 5 s watchdog inactivity limit and to starve the task long enough for the WDT to fire.

Link quality is workable but bursty: RSSI around −65 to −78 dBm at the node (AP sees ~−71 dBm), AP rates hovering near 39 Mbit/s down and 6.5 Mbit/s up (20 MHz), and pings over ~100 s of probing show 0–13 % burst loss with RTT spikes into the seconds. Most of the time throughput is fine (OTA advances ~1 % every 5–15 s for ~1 MiB images), but bursts cause long pauses that later recover - I’ve increased the host‑side per‑chunk timeout in `espota2.py` from 30 s to 600 s, which massively improves reliability without other changes; the remaining failures are due to the device WDT triggering and causing a reboot mid‑transfer.

Because of this transport behavior, the device should not abort after 2.5 s of inactivity. The inactivity budget during the data phase needs to be massively larger - on the order of ≥ 120–180 s by default, ideally `max(10×SRTT, 120 s)` with a hard cap around 600 s - and the OTA task should feed the watchdog while waiting or erasing/programming flash. With that in place there’s no harm in using a 300 s (or even 600 s) ceiling: the connection survives short RTO‑driven stalls and completes once the burst passes, instead of forcing a reboot and guaranteed failure on every non‑ideal update.

We should probably enable SACK to improve recovery from packet loss; clamp the MSS on the host to about 536 B via `TCP_MAXSEG` so more segments are in flight within the ESP's small receive window and fast retransmit can trigger; and consider enabling TCP window scaling to allow a larger receive window so more packets can be in flight, which should also speed up the transfer on marginal links. Until then, longer inactivity thresholds plus watchdog feeding are essential to avoid triggering timeouts all the time on minor wifi hiccups.

bdraco commented:
I increased it from 1s to 2.5s. What do you think it should be set at ?

bdraco committed:
increase to 30s

RubenKelevra commented:
I think your mental model is off: Previously there was no timeout, because the `recv()` was blocking. So if there was no data on the socket, it would just hang, and the effective timeout was 5 seconds.

The 1000 ms was only the timeout for the reading, if there was data on the socket.

But your change made this the effective "no data received within 2.5 seconds" timeout, which was previously 5 seconds - because of the watchdog.

I did 3 update attempts with the default MSS of 14xx bytes and each time the ESP32 rebooted.

I now reduced the MSS to 536 and the first update succeded. However, there's quite a lot of backoff still happening.

Example:

```
968	31.587251099	192.168.8.40	192.168.8.178	TCP	54	[TCP Dup ACK 954#2] 3232 → 52540 [ACK] Seq=44 Ack=38235 Win=5760 Len=0
1207	43.200027800	192.168.8.178	192.168.8.40	TCP	590	[TCP Retransmission] 52540 → 3232 [PSH, ACK] Seq=38235 Ack=44 Win=65349 Len=536
1668	66.752050336	192.168.8.178	192.168.8.40	TCP	322	[TCP Retransmission] 52540 → 3232 [ACK] Seq=38235 Ack=44 Win=65349 Len=268
2838	113.344048911	192.168.8.178	192.168.8.40	TCP	188	[TCP Retransmission] 52540 → 3232 [ACK] Seq=38235 Ack=44 Win=65349 Len=134
```

_Second field is seconds_

This is a typical case of multiple RTO backoffs at the moment (even with clamped down MSS), where you see 46.6 seconds no transmission.

In this case, your new proposed timeout of 30 s could still abort.

_Not sure why the reboot watchdog did not fire here tbh. So maybe the ESP32 was still reading very slowly off the socket? I don't know._

bdraco commented:
I still don't like that it blocks the loop for up to one second just by connecting. 

I think we should fix that before merging

bdraco committed:
do not block until we get first magic byte

bdraco committed:
uint32_t

bdraco committed:
uint32_t

bdraco commented:
@RubenKelevra Thank you for the detailed analysis and packet traces! Your feedback was invaluable in understanding the network timeout scenarios.

Based on your suggestions, I've made the following improvements:

1. **Non-blocking handshake phase**: The initial connection now uses a 10-second timeout to receive the first magic byte in a non-blocking loop, completely eliminating the watchdog reset issue during port scans.

2. **Increased data transfer timeout**: The timeout for read/write operations during data transfer has been increased from 1s to 90s, matching what we use in `aioesphomeapi` for consistency across ESPHome.

3. **Architectural improvements**: Split the OTA handler into `handle_handshake_()` (non-blocking) and `handle_data_()` phases.

The 90s timeout represents an 18x increase from the previous ~5s effective timeout and should handle the vast majority of network conditions. While I understand the case for even longer timeouts, I'd like to keep this PR focused on fixing the blocking/watchdog reset issue. The non-blocking architecture we've implemented fully addresses the core problem.

If we find that 90s is insufficient in practice, we can always revisit the timeout values in a future PR dedicated to network resilience improvements. For now, this aligns with our existing API timeout strategy and solves the immediate issue.

Thanks again for your thorough analysis - it really helped shape this solution!

bdraco committed:
preen

bdraco committed:
Merge remote-tracking branch 'upstream/dev' into ota_fixes

bdraco committed:
adjust

bdraco committed:
preen

bdraco committed:
preen

bdraco committed:
preen

bdraco committed:
preen

bdraco committed:
preen

bdraco committed:
preen

bdraco committed:
preen

bdraco committed:
cleanp

bdraco committed:
preen

bdraco committed:
dry

bdraco committed:
Merge branch 'dev' into ota_fixes

kbx81 commented:
Deployed & working as expected an all 20-something of my ESPs -- original ESP32s, an S2, S3s, C3s and C6s. 👍🏻

RubenKelevra commented:
> @RubenKelevra Thank you for the detailed analysis and packet traces! Your feedback was invaluable in understanding the network timeout scenarios.

You're welcome!

---

I like to highlight the issue we're having here:


With ~1500 Byte packages, we can only send around 3 packages in the buffer on the ESP32, until we end up having to "stop" sending data. The buffer on the receiving end is called receive window.

This receive window will be moved forward (meaning more packages can be pushed onto the network) each time the receiving end removes one package. This "move forward" is communicated to the sending side, which is then allowed to push more data.

With just 3 packages in flight, there's no way to quickly recover from a package loss, because of the way TCP handles loss:

Example:

- We send `#1`
- Receiver is sending an ACKnowledgment back that `#1` was received
- We send `#2`
- ACK back for `#2`
- We send `#3` which is lost
- We send `#4` which is received
- the receiving end sends a "Duplicate ACK" for number `#2'
- We send `#5` which is received
- the receiving end sends a "Duplicate ACK" for number `#2'
- We send `#6` which is received
- the receiving end sends a "Duplicate ACK" for number `#2'

After 3 consecutive "DUP ACKs" the sender will do a "fast recovery" and send `#3` again, as it appears to have been lost. _Note here that TCP actually works with byte ranges, not packets, and everything works concurrently, so the actual ACKs might merged together, if multiple packages are received._

The main issue is now: We need more than 3 packages "in flight" to have fast recovery trigger reliably, as we will otherwise not trigger DUP ACKs quickly.

If the receiver does not receive 3 DUP ACKs, it will fall back to a receive timeout, which means it retry to send after twice the perceived roundtrip of packages, then duplicate this value and retry, then duplicate this value and retry etc.

Meaning, if we can, we always want to trigger "fast recovery" and rely as less often than possible on "RTO", as it may take a lot of time, if the ping is high.

That's why I suggested in https://github.com/esphome/esphome/pull/10164 to reduce the TCP package size, so we can cramp in much more packages on the receiving end, so there are much more chances that we will trigger DUP ACKs.

The other advantage here is, that we can much quicker "recover" from a receiving side reporting "buffer full" as we can fill up if just 500 Byte are removed from the buffer, not 1500 Bytes. This allows us to keep the buffer much fuller on average, and thus reduce the latency between reading from the buffer and refilling, as just 500 bytes reading will trigger the "refill".

The other things we can consider are:

- increasing the receive window by activating dynamically adjusting receive windows, aka "TCP window scaling" in the LWIP stack. This would dynamically adjust the TCP receive window based on the total amount of buffer space left in the package buffer, rather than a fixed ~3 package value in bytes.

- enable SACK. This allows TCP to report back additional package reception (as byte ranges) *past* the lost package, meaning instead of sending "duplicate acks" each time for `#2` in the example LWIP would acknowledge the reception of package `#4`, package `#5` and package `#6`, which then trigger a fast recovery of `#3`.

    The advantage here is, that this does a much quicker recovery if two consecutive packages are lost, as otherwise TCP need to do tripple DUP ACKs to package `#3` and once it has received that one, it need to do triple DUP ACKs for package `#4' - if in the example `#4' was lost too.
    
    Meaning we need at least 6 packages in flight, after two lost packages, to use "fast recovery" for both.

> Based on your suggestions, I've made the following improvements:
> 
>     1. **Non-blocking handshake phase**: The initial connection now uses a 10-second timeout to receive the first magic byte in a non-blocking loop, completely eliminating the watchdog reset issue during port scans.

Given the explanation above, you can probably see that we're not able to do "fast recovery" here. We send a single package, and wait for the receiving end to "acknowledge" that.

Since the receiving end has no idea, that it's supposed to receive something, based on the TCP standard, we will *always* fall back here to the RTO backoff.

Since the only source here for our RTT (round trip time) is the response time to our connection establishment, we need to be quite careful here:

The response time could be quite long, meaning say it takes the ESP32 2 seconds to respond. Then we would send the magic package and after 4 seconds we would wonder "why didn't we get an ACK", and resend the package.

If this package is also lost, we already run into the timeout of 10 seconds, because we double the timeout then to 8 seconds - which gives us a total of 12 seconds before we retry a second time.

Given that we currently have no retry after we selected an "compile and install" in the UI, to reestablish the connection, this will lead to a hard failure of the whole operation, which *could* be seen as frustrating by the user.

I suggest to wait here also quite long, to make sure that we can receive the package properly. 

>     2. **Increased data transfer timeout**: The timeout for read/write operations during data transfer has been increased from 1s to 90s, matching what we use in `aioesphomeapi` for consistency across ESPHome.

Fair compromise.
 
> The 90s timeout represents an 18x increase from the previous ~5s effective timeout and should handle the vast majority of network conditions. While I understand the case for even longer timeouts, I'd like to keep this PR focused on fixing the blocking/watchdog reset issue. The non-blocking architecture we've implemented fully addresses the core problem.

As we're already deep into a technical discussion at this point, I like to point out one of the major issues we're currently having with the OTA, which is causing most of the issues we're seeing:

We currently send 8 KByte blocks. Wait for a return of an "OK I GOT THAT" and then send another 8 Byte block. Then intent here seems to be, that we're needing a progress bar which reflects the actual progress.

But sending 8 KByte blocks and then wait, gives us the following issue:

At the end of each block we run into the issue, that TCP cannot do a fast recovery and instead needs to rely on RTO backoff to submit the missing packages.

Instead, I suggest we redesign the system to instead send a notification out on from the receiving end **asynchronously** that 8 KByte has been received, and this is then reflected as an update of the progress.

This means we will send the whole stream in one go, and update the progress based on "I've got 1 KByte block" updates from the Microcontroller and let TCP handle the rest, as it's designed to do.

This would allow TCP to act much better in terms of loss recovery and should also speed up the update process, as we don't "stall" at the end of each 8 KByte block with some chance on any type of package loss.


> If we find that 90s is insufficient in practice, we can always revisit the timeout values in a future PR dedicated to network resilience improvements. For now, this aligns with our existing API timeout strategy and solves the immediate issue.


Good approach! :)

RubenKelevra commented:
> That said, anecdotally, I feel like I've heard about a lot of issues with the supermini modules in general as it relates to RF performance.

Yeah, I mean they work 90% of the time fine, I got like 7 of them in operation, only two got issues, which I attribute to:

1) Being quite far away from the AP
2) Being near metal objects. One is close to a metal radiator and a bath tube and the other one is in an enclosure which got some metal paint on the outside (just 50% around it) and it's near a bar, which got a metal sheet in the front.

So overall *not* ideal RF conditions, as it probably sees a lot of reflections. But still the day to day operation doesn't really seem to be affected, I just see issues doing the OTA.

RubenKelevra commented:
> As we're already deep into a technical discussion at this point, I like to point out one of the major issues we're currently having with the OTA, which is causing most of the issues we're seeing:
> 
> We currently send 8 KByte blocks. Wait for a return of an "OK I GOT THAT" and then send another 8 Byte block. Then intent here seems to be, that we're needing a progress bar which reflects the actual progress.
> 
> But sending 8 KByte blocks and then wait, gives us the following issue:
> 
> At the end of each block we run into the issue, that TCP cannot do a fast recovery and instead needs to rely on RTO backoff to submit the missing packages.
> 
> Instead, I suggest we redesign the system to instead send a notification out on from the receiving end **asynchronously** that 8 KByte has been received, and this is then reflected as an update of the progress.
> 
> This means we will send the whole stream in one go, and update the progress based on "I've got 1 KByte block" updates from the Microcontroller and let TCP handle the rest, as it's designed to do.
> 
> This would allow TCP to act much better in terms of loss recovery and should also speed up the update process, as we don't "stall" at the end of each 8 KByte block with some chance on any type of package loss.

Wrote an implementation of that, will send a PR shortly, when I completed the testing.

bdraco commented:
RubenKelevra kbx81: Thank you both for the excellent feedback and thorough testing!

## Current Status

I'm happy with this PR at this point. After testing on 40 different devices and with @kbx81's additional testing (thank you Keith!), I'm confident in the current implementation. 

With beta releasing in less than 48 hours, I want to be very careful not to expand the scope of this PR. The current changes:
- Fix the watchdog reset during port scanning
- Fix the watchdog reset when OTA stalls for more than ~5 seconds
- Implement proper non-blocking socket handling with reasonable timeouts

These changes address the critical issues while maintaining stability.

## Network Resilience Improvements

RubenKelevra: Your technical analysis of TCP performance and packet recovery has been incredibly valuable. The insights keep evolving as we learn more about the network behavior. I'd love to see you continue exploring these network resilience improvements in follow-up PRs. 

More analysis is needed as the picture keeps changing, but your deep dive into the TCP stack behavior is exactly the kind of investigation that will help make ESPHome more robust on challenging networks.

## OTA Block Size

Regarding the 8KB block size - you're right that it could be optimized. The current value was chosen to maximize throughput on ethernet-connected devices and solid WiFi connections. Dropping it to something that fits under the MTU (probably 1360 bytes) would be a safe change that could improve reliability on marginal connections.

I'd be OK with changing just the block size in a follow-up PR before beta, as that's easy to test and verify (and we've changed it before without issues). However, any other network optimizations should be reserved for 2025.9.x so they have a full month to go through dev nightlies.

## Moving Forward

Let's get this critical fix merged for beta, and then we can work on the network performance improvements in subsequent PRs. This approach ensures:
1. Users get immediate relief from the watchdog reset issues
2. We maintain stability for the beta release
3. Network optimizations get proper testing time

Thanks again for all your valuable input and collaboration!
